lib/shared/bucketing-assembly-script/assembly/types/featureV2.ts

```diff
@@ -0,0 +1,124 @@
+import { JSON } from '@devcycle/assemblyscript-json/assembly'
+import {
+    getJSONArrayFromJSON,
+    getJSONObjFromJSON, getJSONObjFromJSONOptional, getJSONValueFromJSON,
+    getStringFromJSON,
+    isValidString,
+    jsonArrFromValueArray
+} from '../helpers/jsonHelpers'
+import { FeatureConfigurationV2 } from './featureConfigurationV2'
+
+const validTypes = ['release', 'experiment', 'permission', 'ops']
+
+export class FeatureV2 extends JSON.Value {
+    readonly _id: string
+    readonly type: string
+    readonly key: string
+    readonly variations: Variation[]
+    readonly configuration: FeatureConfigurationV2
+    readonly settings: JSON.Obj | null
+
+    private readonly _variationsById: Map<string, Variation>
+
+    constructor(feature: JSON.Obj) {
+        super()
+        this._id = getStringFromJSON(feature, '_id')
+
+        this.type = isValidString(feature, 'type', validTypes)
+
+        this.key = getStringFromJSON(feature, 'key')
+
+        const variationsJSON = getJSONArrayFromJSON(feature, 'variations').valueOf()
+        const variations = new Array<Variation>()
+        const variationsById = new Map<string, Variation>()
+        for (let i = 0; i < variationsJSON.length; i++) {
+            const variation = new Variation(variationsJSON[i] as JSON.Obj)
+            variations.push(variation)
+            variationsById.set(variation._id, variation)
+        }
+        this.variations = variations
+        this._variationsById = variationsById
+
+        this.configuration = new FeatureConfigurationV2(getJSONObjFromJSON(feature, 'configuration'))
+
+        this.settings = getJSONObjFromJSONOptional(feature, 'settings')
+    }
+
+    getVariationById(variationId: string): Variation | null {
+        if (!this._variationsById.has(variationId)) return null
+        return this._variationsById.get(variationId)
+    }
+
+    stringify(): string {
+        const json = new JSON.Obj()
+        json.set('_id', this._id)
+        json.set('type', this.type)
+        json.set('key', this.key)
+        json.set('variations', jsonArrFromValueArray(this.variations))
+        json.set('configuration', this.configuration)
+        return json.stringify()
+    }
+}
+
+export class Variation extends JSON.Value {
```

Comment:
can this not be imported from v1?

Reply:
yeah, but then we need to be importing different classes from v1/v2 depending on the class. Unless you're proposing that I import the classes from v1 and then re-export them in v2

Reply:
That could work but it will make removing v1 more of a pain

Reply:
yeah thats fair, I think this just makes it harder to maintain two versions but I see your point about making deprecating v1 easier. 

would it be easier to have 2 commits, 1 copying over v1 into v2 and the other making the changes? if not i think we can get by with this since its not a lot of updates being made

Reply:
that would have been nicer, how about I go through and finish commenting on the changes between v1 and v2 so its easier to read?

Reply:
Added comments to all of the changes made to duplicated v1 files, hopefully makes reviewing easier